File: 11.盛最多水的容器.cpp
/*
 * @lc app=leetcode.cn id=11 lang=cpp
 *
 * [11] 盛最多水的容器
 */
class Solution {
public:
    int maxArea(vector<int>& height) {
        // 思路：双指针法。
        int maxarea = 0, l = 0, r = height.size() - 1;
        while (l < r)
        {
            if (height[l] <= height[r])
            {
                maxarea = max(maxarea, (r - l) * height[l]);
                ++l;
            }
            else if (height[l] > height[r]) 
            {
                maxarea = max(maxarea, (r - l) * height[r]);
                --r;
            }
        }
        return maxarea;       
    }
};

File: 11.盛最多水的容器.go
/*
 * @lc app=leetcode.cn id=11 lang=golang
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
func maxArea(height []int) int {
  var (
    left = 0
    right = len(height) - 1
    res = 0
  )

  for left < right {
    width := right - left
    high := 0
    if height[left] < height[right] {
      high = height[left]
      left++
    } else {
      high = height[right]
      right--
    }
    res = max(res, width * high)
  }

  return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

